clr/commands.py

```diff
@@ -135,45 +131,149 @@ class Namespace:
+                    raise AssertionError(f'Unexpected arg type for {param.name} in {command_name}: '
+                                         f'{default_type}')
+
+                # Put the default in the help text to clarify behavior when it is not specified.
+                help_text = f"Optional. Defaults to {param.name}='{param.default}'."
+
+                if default_type == bool:
+                    # Add both the --arg and --noarg options, but make them mutally exclusive.
+                    group = parser.add_mutually_exclusive_group()
+                    group.add_argument(f'--{param.name}', action='store_true', help=help_text)
+                    group.add_argument(f'--no{param.name}', dest=param.name, action='store_false',
+                        help=help_text)
+                else:
+                    # Add both as optional (nargs=?) positional and named (--arg) for flexibility.
+                    # Mutually exclusive and have the same dest.
+                    group = parser.add_mutually_exclusive_group()
+                    group.add_argument(param.name, nargs='?', type=default_type,
+                        help=f'{help_text} Can also be specified with --{param.name}.')
+                    group.add_argument(f'--{param.name}', type=default_type,
+                        help=f'{help_text} Can also be specified with positional arg {param.name}.')
+        return parser
+
+
 @dataclass
 class ErrorLoadingNamespace:
-    """Psuedo namespace for when one can't be loaded to show the error message."""
+    """Psuedo namespace for when one can't be loaded to show the error message in `clr help`."""
     key: str
     error: Exception
 
-    # Satisfy the same properties of a `Namespace`, but never have any actual
-    # commands.
-    commands = frozenset()
-    command_specs = {}
+    @property
+    def commands(self):
+        return tuple()
+
+    @property
+    def command_specs(self):
+        return {}
 
     @property
     def descr(self):
         return f"ERROR Could not load. See `clr help {self.key}`"
 
     @property
     def longdescr(self):
-        return f"""Error importing module '{clr.config.commands()[self.key]}' for namespace '{self.key}':\n\n{self.error}"""
+        return (f"Error importing module '{NAMESPACE_MODULE_PATHS[self.key]}' for namespace "
+                f"'{self.key}':\n\n{self.error}")
 
 @dataclass(frozen=True)
 class NamespaceCacheEntry:
+    """Picke-able subset of Namespace for NamespaceCache."""
+    key: str
     descr: str
     longdescr: str
     command_specs: dict
 
     @staticmethod
     def create(namespace):
-        return NamespaceCacheEntry(namespace.descr, namespace.longdescr, namespace.command_specs)
-
-    @property
-    def commands(self):
-        return sorted(self.command_specs.keys())
+        return NamespaceCacheEntry(namespace.key, namespace.descr,
+            namespace.longdescr, namespace.command_specs)
+# Steal some functionality.
```

Comment:
Hm, definitely a somewhat strange way to do polymorphism. =)

Maybe these methods should just be intentionally general?

Reply:
Just for code organization I wanted to keep them nested. And it's nice to be able to call `namespace.commands`.

Because they are `@dataclass`es I can't do regular inheritance without it complaining